New Command: Get-CrmSPDocLoc

RichieRoxx opened the pull request:
Hi Arjan, as announced via linked in: I've added a new cmdlets Get-CRMSPDocLoc to easily retrieve the Urls of a SharePoint Document Location. I work a lot with CRM and SharePoint-Integrations and regularly make use of this. Happy to adapt anything so you can pull this into your repo. Looking forward to hearing from you.

Commit:
Introduction of new Admininstration Cmdlet get-crmspdocloc

this cmdlet enabled you to pass the recordid of a docloc and it will return the absolute URL of the corresponding sharepoint-folder. Further it accepts the recordid of any entity to record the according absolute-urls of all related doclocs...

RichieRoxx committed:
Delete storage.ide-shm

RichieRoxx committed:
Delete storage.ide

RichieRoxx committed:
Delete storage.ide-wal

RichieRoxx commented:
I noticed I need to switch codesigning back on. can certainly do that, but would like to collect the whole of your feedback first.

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+{
+    [Cmdlet(VerbsCommon.Get, "CrmSPDocLoc", HelpUri = HelpUrlConstants.GetRoleHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
+    [OutputType(typeof(Entity))]
+    public class GetSPDocLoc :CrmCmdlet
```
Use fullname for Classname and all Cmdlet classes end in 'Command'
Inherit from CrmOrganizationCmdlet

```suggestion
    public class GetSPDocucmentLocationCommand :CrmOrganizationCmdlet
```

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+
+namespace AMSoftware.Crm.PowerShell.Commands.Administration
+{
+    [Cmdlet(VerbsCommon.Get, "CrmSPDocLoc", HelpUri = HelpUrlConstants.GetRoleHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
```
I prefer fullnames for my Cmdlets. Autocomplete in PowerShell will help with typing, and this way the comdlets are clear.
Add a HelpUrl in the HelpUrlConstants, or leave out if no help provided.
```suggestion
    [Cmdlet(VerbsCommon.Get, "CrmSPDocumentLocation", HelpUri = HelpUrlConstants.GetSPDocumentLocationHelpUrl, SupportsPaging = true, DefaultParameterSetName = "")]
```

RichieRoxx replied:
accepted

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+    [OutputType(typeof(Entity))]
+    public class GetSPDocLoc :CrmCmdlet
+    {
+        [Parameter(Mandatory = true)]
```
Include validation on the parameter if it is mandatory.
```suggestion
        [Parameter(Mandatory = true)]
        [ValidateNotNullOrEmpty]
```

RichieRoxx replied:
accepted

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+        [Parameter(Mandatory = true)]
+        public Guid[] Id { get; set; }
+
+        [ArgumentCompleter(typeof(EntityArgumentCompleter))]
+        [Parameter]
+        public string Entity { get; set; }
```
If not using this then remove. 
I would prefer to keep the parameter and add an additional one for EntityId (Guid[]) so one or more specific Records can be provided. And use the Id parameter in a separate ParameterSet to retrieve a specific DocumentLocation record.

RichieRoxx replied:
accepted

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+
+        private void RetrieveUrlById(Guid id)
+        {
+            var sourceUrlRequest = new RetrieveAbsoluteAndSiteCollectionUrlRequest
```
Prefer generic OrganizationRequest to limit the need for the Messages namespace. See SetOwnerCommand.cs for example.
```suggestion
            OrganizationRequest sourceUrlRequest = new OrganizationRequest("RetrieveAbsoluteAndSiteCollectionUrl")
```

RichieRoxx replied:
In general I'd prefer strongly typed requests over generic requests as you suggest. Could you please elaborate on why you'd rather not want to use the messages namespace? Thank you!

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+            {
+                Target = new EntityReference("sharepointdocumentlocation", id)
+            };
+            var sourceUrlResponse = (RetrieveAbsoluteAndSiteCollectionUrlResponse)CrmContext.OrganizationProxy.Execute(sourceUrlRequest);
```
Use generic OrganizationResponse to limit the need for the Messages namespace. See GetLanguageCommand
And use the Repository to execute the request
```suggestion
            OrganizationResponse sourceUrlResponse = _repository.Execute(sourceUrlRequest);
```

AMSoftwareNL commented on AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs:
```diff
@@ -0,0 +1,65 @@
+            {
+                foreach (var id in Id)
+                {
+                    QueryByAttribute query = new QueryByAttribute("sharepointdocumentlocation");
+                    query.AddAttributeValue("regardingobjectid", id);
```
Include Entity and EntityId in query. There won't be duplicate ID's but regarding is a lookup so includes the Entity, and is an extra check for the user.

RichieRoxx replied:
not exactly sure what you mean by this. Please see the latest changes and check if this is already obsolete.

AMSoftwareNL reviewed (commented):
Good addition to the library. Made some remarks and suggestions to match with the existing commands.
Would suggest to make two ParameterSets. One to get the document location by ID, and one to get by Record (Entity and EntityId as separate parameters).

AMSoftwareNL commented:
You can disable code signing to compile, test and the pull request. I have the key only in my local repository, and will add on build of a new release. It should all work without the assembly being signed. It's just a quality thing.

RichieRoxx committed:
Update AMSoftware.Crm.PowerShell.Commands/Administration/GetSPDocLoc.cs

Co-authored-by: Arjan Meskers <AMSoftwareNL@users.noreply.github.com>

Commit:
added parameter validation
added outputtype

Commit:
Renaming Class and Command

Commit:
- refactored to two parametersets
- removed Entity-Parameter
- renamed Verb as suggested
- added documentation skeleton file
